File: 2026 05월/5월 4주차/JeongEon8/01/[PGS - Lv2] 할인 행사.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    
    // 1. 맵에 {물품이름, 개수}로 10개 미리 넣는다.
    map<string, int> m;
    for(int i = 0; i < 10; i++){
        m[discount[i]]++;
    }
    
    // 2. 사려고 했던 물품의 개수가 맞는지 확인한다.
    for(int i = 0; i < discount.size() - 9; i++){
        bool signUp = true;
        
        for(int j = 0; j < want.size(); j++){
            if(m[want[j]] != number[j]){
                signUp = false;
                break;
            }
        }
        
        if(signUp){
            answer++;
        }
        
        m[discount[i]]--;
        if(i < discount.size() - 10){
            m[discount[i+10]]++;
        }
    }
    
    return answer;
}

File: 8월 1주차/hyuksanglee/03/[PGS - Lv2] 할인 행사.java
import java.util.*;
class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Queue<String> que = new ArrayDeque<>();
        Map<String, Integer> map = new HashMap<>();
        for(int i = 0; i<want.length; i++){
            map.put(want[i], number[i]);
        }
        int total = 10;
        for(int i = 0; i<10; i++){
            String prd = discount[i];
            que.add(prd);
            if (map.containsKey(prd)) {
                int nextCount = map.get(prd) - 1;
                map.put(prd, nextCount);
                
                if (nextCount >= 0) {
                    total -= 1;
                }
            }
        }
        if (total == 0) {
                answer++;
            }
        
        for(int i = 10; i<discount.length; i++){
            
            String bfPrd = que.poll();
            if (map.containsKey(bfPrd)) {
                int nextCount = map.get(bfPrd) + 1;
                map.put(bfPrd, nextCount);

                if (nextCount > 0) {
                    total += 1;
                }
            }
            
            String prd = discount[i];
            que.add(prd);
            if (map.containsKey(prd)) {
                int nextCount = map.get(prd) - 1;
                map.put(prd, nextCount);
                

                if (nextCount >= 0) {
                    total -= 1;
                }
            }
            if (total == 0) {
                answer++;
            }
        }
        
        return answer;
    }
}
